tests/unit-tests/scene/test_surface_stack.cpp

```diff
@@ -1374,3 +1374,99 @@ TEST_F(SurfaceStack, when_screen_is_locked_surface_ignores_surface)
+    EXPECT_THAT(
+        stack.scene_elements_for(compositor_id),
+        ElementsAre(
+            SceneElementForStream(stub_buffer_stream2),
+            SceneElementForStream(stub_buffer_stream3),
+            SceneElementForStream(stub_buffer_stream1)));
+
+    Mock::VerifyAndClearExpectations(&observer);
+}
+
+TEST_F(SurfaceStack, single_surfaces_can_be_swapped)
+{
+    stack.add_surface(stub_surface1, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface2, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface3, mi::InputReceptionMode::normal);
+
+    NiceMock<MockSceneObserver> observer;
+    stack.add_observer(mt::fake_shared(observer));
+
+    EXPECT_CALL(
+        observer,
+        surfaces_reordered(UnorderedElementsAre(LockedEq(stub_surface1))))
+        .Times(1);
+
+    EXPECT_CALL(
+        observer,
+        surfaces_reordered(UnorderedElementsAre(LockedEq(stub_surface2))))
+        .Times(1);
+
+    stack.swap_z_order({stub_surface1}, {stub_surface2});
+    EXPECT_THAT(
+        stack.scene_elements_for(compositor_id),
+        ElementsAre(
+            SceneElementForStream(stub_buffer_stream2),
+            SceneElementForStream(stub_buffer_stream1),
+            SceneElementForStream(stub_buffer_stream3)));
+
+    Mock::VerifyAndClearExpectations(&observer);
+}
+
+TEST_F(SurfaceStack, multiple_surfaces_can_be_swapped)
+{
+    stack.add_surface(stub_surface1, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface2, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface3, mi::InputReceptionMode::normal);
+
+    NiceMock<MockSceneObserver> observer;
+    stack.add_observer(mt::fake_shared(observer));
+
+    EXPECT_CALL(
+        observer,
+        surfaces_reordered(UnorderedElementsAre(LockedEq(stub_surface1), LockedEq(stub_surface2))))
+        .Times(1);
+
+    EXPECT_CALL(
+        observer,
+        surfaces_reordered(UnorderedElementsAre(LockedEq(stub_surface3))))
+        .Times(1);
+
+    stack.swap_z_order({stub_surface1, stub_surface2}, {stub_surface3});
+    EXPECT_THAT(
+        stack.scene_elements_for(compositor_id),
+        ElementsAre(
+            SceneElementForStream(stub_buffer_stream3),
+            SceneElementForStream(stub_buffer_stream1),
+            SceneElementForStream(stub_buffer_stream2)));
+
+    Mock::VerifyAndClearExpectations(&observer);
+}
+
+TEST_F(SurfaceStack, swapping_overlapping_surface_sets_causes_an_exception)
+{
+    stack.add_surface(stub_surface1, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface2, mi::InputReceptionMode::normal);
+    stack.add_surface(stub_surface3, mi::InputReceptionMode::normal);
+
+    EXPECT_THAT([&]() { stack.swap_z_order({stub_surface1, stub_surface2}, {stub_surface2, stub_surface3}); },
+                Throws<std::logic_error>(Property(&std::logic_error::what,
+                                                    HasSubstr("Swapped sets must not have any overlap"))));
+}
```

Comment:
Throwing exceptions is not normally part of the contract. Particularly for precondition violations. (Preconditions are the responsibility of caller not the callee.)

Exceptions should be used to report that the post-condition cannot be met.

So, I wouldn't have this test because, assuming the precondition is met, then it doesn't happen.

But if I did, then I'd also want this to pass:

```c++
TEST_F(SurfaceStack, swapping_overlapping_surface_sets_causes_an_exception2)
{
    stack.add_surface(stub_surface1, mi::InputReceptionMode::normal);
    stack.add_surface(stub_surface3, mi::InputReceptionMode::normal);

    EXPECT_THAT([&]() { stack.swap_z_order({stub_surface1, stub_surface2}, {stub_surface2, stub_surface3}); },
                Throws<std::logic_error>(Property(&std::logic_error::what,
                                                  HasSubstr("Swapped sets must not have any overlap"))));
}
```

Reply:
I have moved the check to BasicWindowManager instead :+1: